backend/corpora/common/entities/dataset.py

```diff
@@ -1,9 +1,106 @@
+        publication_doi: str = "",
+        artifacts: list = None,
+        contributors: list = None,
+        deployment_directories: list = None,
+    ) -> "Dataset":
+        """
+        Creates a new dataset and related objects
+        :param revision:
+        :param name:
+        :param organism:
+        :param organism_ontology:
+        :param tissue:
+        :param tissue_ontology:
+        :param assay:
+        :param assay_ontology:
+        :param disease:
+        :param disease_ontology:
+        :param sex:
+        :param ethnicity:
+        :param ethnicity_ontology:
+        :param source_data_location:
+        :param preprint_doi:
+        :param publication_doi:
+        :param artifacts:
+        :param contributors:
+        :param deployment_directories:
+        :return:
+        """
+        uuid = cls.db.generate_id(DbDataset)
+
+        """
+        Prevent accidentally linking an existing row to a different Dataset. This maintains the relationship of one
+        to many for artifacts and deployment_directories
+        """
+        artifacts = artifacts if artifacts else []
+        deployment_directories = deployment_directories if deployment_directories else []
+        contributors = contributors if contributors else []
+
+        [artifact.pop("id", None) for artifact in artifacts]  # sanitize of ids
+        [deployment_directory.pop("id", None) for deployment_directory in deployment_directories]  # sanitize of ids
+
+        new_db_object = DbDataset(
+            id=uuid,
+            revision=revision,
+            name=name,
+            organism=organism,
+            organism_ontology=organism_ontology,
+            tissue=tissue,
+            tissue_ontology=tissue_ontology,
+            assay=assay,
+            assay_ontology=assay_ontology,
+            disease=disease,
+            disease_ontology=disease_ontology,
+            sex=sex,
+            ethnicity=ethnicity,
+            ethnicity_ontology=ethnicity_ontology,
+            source_data_location=source_data_location,
+            preprint_doi=preprint_doi,
+            publication_doi=publication_doi,
+            artifacts=cls._create_sub_objects(artifacts, DbDatasetArtifact, dataset_id=uuid),
+            deployment_directories=cls._create_sub_objects(
+                deployment_directories, DbDeploymentDirectory, dataset_id=uuid
+            ),
+        )
+
+        #  Linking many contributors to many datasets
+        contributors = cls._create_sub_objects(contributors, DbContributor)
+        contributor_dataset_ids = [dict(contributor_id=contributor.id, dataset_id=uuid) for contributor in contributors]
+        dataset_contributor = cls._create_sub_objects(contributor_dataset_ids, DbDatasetContributor)
+
+        cls.db.session.add(new_db_object)
+        cls.db.session.add_all(contributors)
+        cls.db.session.commit()
+
+        cls.db.session.add_all(dataset_contributor)
+        cls.db.session.commit()
```

Comment:
You might want to catch exceptions here if the commit fails and log the errors as well as rollback the transaction.

Reply:
added